File: python/nexcore/http.py
"""底层 HTTP 传输层.

各业务命名空间(payment / exchange / energy / smtp)调用本类发请求,
不直接接触 requests,保持 namespace 关注业务逻辑.
"""

from __future__ import annotations
import json
from typing import Any, Dict, Optional

try:
    import requests
except ImportError as e:
    raise ImportError(
        "nexcore SDK 需要 `requests`,请运行:pip install requests"
    ) from e

from .errors import NexCoreError


class Http:
    """封装 requests 调用,处理 envelope 解包 + 异常统一."""

    def __init__(
        self,
        base_url: str,
        timeout: int = 30,
        verify_ssl: bool = True,
        user_agent: str = "NexCore-Python-SDK/3.0.0",
    ):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.verify_ssl = verify_ssl
        self.user_agent = user_agent

    def request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        query: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """发送 HTTP 请求.

        Args:
            method: HTTP 方法 GET / POST / PUT / DELETE.
            path: 以 / 开头的路径,如 ``/api/v1/pay/create``.
            body: JSON 请求体(自动序列化).
            query: query 参数.
            headers: 额外 header.

        Returns:
            响应中的 ``data`` 段(已解 ``{code, message, data}`` envelope).

        Raises:
            NexCoreError: 网络错误 / HTTP 错误 / 业务 code != 0.
        """
        url = self.base_url + path
        h: Dict[str, str] = {
            "Accept": "application/json",
            "User-Agent": self.user_agent,
        }
        if body is not None:
            h["Content-Type"] = "application/json"
        if headers:
            h.update(headers)

        kwargs: Dict[str, Any] = {
            "headers": h,
            "timeout": self.timeout,
            "verify": self.verify_ssl,
        }
        if query:
            # 过滤空值
            kwargs["params"] = {k: v for k, v in query.items() if v not in (None, "")}
        if body is not None:
            kwargs["data"] = json.dumps(body, ensure_ascii=False).encode("utf-8")

        try:
            resp = requests.request(method.upper(), url, **kwargs)
        except requests.RequestException as e:
            raise NexCoreError(f"HTTP request failed: {e}") from e

        request_id = resp.headers.get("X-Trace-Id")
        try:
            data = resp.json()
        except ValueError:
            raise NexCoreError(
                f"HTTP {resp.status_code}: {resp.text[:200]}",
                http_status=resp.status_code,
                request_id=request_id,
            )

        if resp.status_code >= 400 or not isinstance(data, dict):
            msg = data.get("message", f"HTTP {resp.status_code}") if isinstance(data, dict) else str(data)
            code = data.get("code", -1) if isinstance(data, dict) else -1
            raise NexCoreError(msg, code=code, http_status=resp.status_code, request_id=request_id)

        if data.get("code") not in (0, None):
            raise NexCoreError(
                data.get("message", "unknown error"),
                code=data.get("code", -1),
                http_status=resp.status_code,
                request_id=request_id,
            )

        return data.get("data", data)

File: node/src/http.js
'use strict';

/**
 * NexCore SDK 底层 HTTP 传输层.
 *
 * 仅依赖 Node.js 内置 http/https 模块,零外部运行时依赖.
 * 业务命名空间(payment / exchange / energy / smtp)调用本类发请求,
 * 不直接接触 http,保持各 namespace 关注业务逻辑.
 */

const http = require('http');
const https = require('https');
const { URL } = require('url');

const { NexCoreError } = require('./errors');

class Http {
  /**
   * @param {object} opts
   * @param {string} opts.baseUrl
   * @param {number} [opts.timeout=30000]
   * @param {string} [opts.userAgent]
   */
  constructor({ baseUrl, timeout = 30000, userAgent = 'NexCore-Node-SDK/3.0.0' }) {
    if (!baseUrl) throw new NexCoreError('baseUrl is required');
    this.baseUrl = String(baseUrl).replace(/\/+$/, '');
    this.timeout = timeout;
    this.userAgent = userAgent;
  }

  /**
   * 发送 HTTP 请求.
   *
   * @param {string} method - HTTP 方法 GET / POST / PUT / DELETE
   * @param {string} path - 以 / 开头的路径,如 "/api/v1/pay/create"
   * @param {object} [opts]
   * @param {object|string} [opts.body] - JSON body(自动序列化)
   * @param {Object<string,string|number>} [opts.query] - query 参数
   * @param {Object<string,string>} [opts.headers] - 额外 header
   * @returns {Promise<object>} 响应中的 data 段(已解 {code,message,data} envelope)
   * @throws {NexCoreError}
   */
  request(method, path, opts = {}) {
    const { body, query, headers } = opts;
    const url = new URL(this.baseUrl + path);

    if (query && typeof query === 'object') {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null && v !== '') {
          url.searchParams.set(k, String(v));
        }
      }
    }

    const h = {
      Accept: 'application/json',
      'User-Agent': this.userAgent,
      ...(headers || {}),
    };

    let payload;
    if (body !== undefined && body !== null) {
      payload = typeof body === 'string' ? body : JSON.stringify(body);
      h['Content-Type'] = 'application/json';
      h['Content-Length'] = Buffer.byteLength(payload);
    }

    const lib = url.protocol === 'https:' ? https : http;
    const reqOpts = {
      method: method.toUpperCase(),
      hostname: url.hostname,
      port: url.port || (url.protocol === 'https:' ? 443 : 80),
      path: url.pathname + url.search,
      headers: h,
      timeout: this.timeout,
    };

    return new Promise((resolve, reject) => {
      const req = lib.request(reqOpts, (res) => {
        const chunks = [];
        res.on('data', (c) => chunks.push(c));
        res.on('end', () => {
          const text = Buffer.concat(chunks).toString('utf8');
          const requestId = res.headers['x-trace-id'] || null;
          let json;
          try {
            json = JSON.parse(text);
          } catch {
            return reject(new NexCoreError(
              `HTTP ${res.statusCode}: ${text.slice(0, 200)}`,
              -1, requestId, res.statusCode,
            ));
          }
          if (res.statusCode >= 400 || (json.code !== undefined && json.code !== 0)) {
            return reject(new NexCoreError(
              json.message || `HTTP ${res.statusCode}`,
              json.code ?? -1, requestId, res.statusCode,
            ));
          }
          resolve(json.data !== undefined ? json.data : json);
        });
      });
      req.on('timeout', () => {
        req.destroy();
        reject(new NexCoreError('request timeout', -1));
      });
      req.on('error', (e) => reject(new NexCoreError(`HTTP request failed: ${e.message}`, -1)));
      if (payload) req.write(payload);
      req.end();
    });
  }
}

module.exports = { Http };
